File: 4_Median_of_Two_Sorted_Arrays/4_Median_of_Two_Sorted_Arrays.c
/**
 * @author CodeamonCat/CodeamonCat@gmail.com
 */

double findMedianSortedArrays(int* nums1, int nums1Size, int* nums2,
                              int nums2Size) {
    // nums1Size <= nums2Size is required
    if (nums1Size > nums2Size) {
        return findMedianSortedArrays(nums2, nums2Size, nums1, nums1Size);
    }

    int median = (nums1Size + nums2Size + 1) / 2;
    int left = 0;
    int right = nums1Size;

    while (left < right) {
        int mid = left + (right - left) / 2;
        int mid2 = median - mid;

        if (nums1[mid] < nums2[mid2 - 1]) {
            left = mid + 1;
        } else {
            right = mid;
        }
    }

    int mid = left;
    int mid2 = median - left;

    int nums1_left_max = mid == 0 ? INT_MIN : nums1[mid - 1];
    int nums2_left_max = mid2 == 0 ? INT_MIN : nums2[mid2 - 1];
    int nums1_right_min = mid == nums1Size ? INT_MAX : nums1[mid];
    int nums2_right_min = mid2 == nums2Size ? INT_MAX : nums2[mid2];

    if ((nums1Size + nums2Size) % 2 == 1) {
        return fmax(nums1_left_max, nums2_left_max);
    } else {
        return (fmax(nums1_left_max, nums2_left_max) +
                fmin(nums1_right_min, nums2_right_min)) /
               2.0;
    }
}

File: 4_Median_of_Two_Sorted_Arrays/4_Median_of_Two_Sorted_Arrays.py
__author__ = "CodeamonCat"
__email__ = "CodeamonCat@gmail.com"


class Solution:

    def findMedianSortedArrays(self, nums1: List[int],
                               nums2: List[int]) -> float:

        # len(nums1) <= len(nums2) is required
        if (len(nums1) > len(nums2)):
            nums1, nums2 = nums2, nums1

        median = (len(nums1) + len(nums2) + 1) // 2
        left = 0
        right = len(nums1)

        while left < right:
            mid = left + (right - left) // 2
            mid2 = median - mid

            if nums1[mid] < nums2[mid2 - 1]:
                left = mid + 1
            else:
                right = mid

        mid = left
        mid2 = median - mid

        nums1_left_max = float('-inf') if mid == 0 else nums1[mid - 1]
        nums1_right_min = float('inf') if mid == len(nums1) else nums1[mid]
        nums2_left_max = float('-inf') if mid2 == 0 else nums2[mid2 - 1]
        nums2_right_min = float('inf') if mid2 == len(nums2) else nums2[mid2]

        if (len(nums1) + len(nums2)) % 2 == 0:
            return (max(nums1_left_max, nums2_left_max) +
                    min(nums1_right_min, nums2_right_min)) / 2
        else:
            return max(nums1_left_max, nums2_left_max)
